add documentation for redline QPS feature in OSB (#9864)
opensearch-project/documentation-website: _benchmark/reference/commands/redline-test.md

---
layout: default
title: redline-test
nav_order: 85
parent: Command reference
grand_parent: OpenSearch Benchmark Reference
redirect_from: 
  - /benchmark/commands/redline-test/
---

# Redline Testing
The `--redline-test` command enables OpenSearch Benchmark to automatically determine the maximum request throughput your OpenSearch cluster
can handle under increasing load. This dynamically adjusts the number of active clients based on real-time cluster performance, and can
assist with capacity planning or detecting performance regressions.

### Usage
To perform a redline test, use the `execute-test` command with the `--redline-test` flag, using a timed test procedure.

> An example test procedure would be:

Reviewer: Nit: 
> Example of a timed test procedure

       "clients": {{ search_clients | default(20) }}
    }
  ]
}
```

which allows us to use:
```bash
opensearch-benchmark execute-test \
  --pipeline=benchmark-only \
  --target-hosts=<your-opensearch-cluster> \
  --workload=<workload> \
  --test-procedure=timed-mode-test-procedure \
  --redline-test
```

### How It Works

When the --redline-test flag is used:

1. Client Initialization: OSB spawns a large number of clients (default is 1000, configurable via `--redline-test=<int>`).

2. Feedback Mechanism: OSB will begin ramping up the number of active clients. A FeedbackActor monitors real-time request failures and adjusts the number of active clients accordingly.

3. Shared State Coordination: OSB uses Python's multiprocessing library to manage shared dictionaries and queues for inter-process communication:
    - Workers: Create and share client state maps with the WorkerCoordinatorActor.
    - WorkerCoordinatorActor: Aggregates and forwards client states to the FeedbackActor.
    - FeedbackActor: Scales up clients until request errors are detected, then pauses clients and enters a sleep period before resuming.

> 

Reviewer: Might be good to include a simplified diagram of the one from your RFC https://github.com/opensearch-project/opensearch-benchmark/issues/785

Reviewer: The simplified diagram can emphasize the components above (Workers, WorkerCoordinatorActor, and Feedback Actor) so that it's clear to the user what's new and what's happening

Author: Added a simplified image to the page. Lmk if it makes sense to you?

### Additional Options
- The number of clients to scale up can be customized using `--redline-test=<int>`. For example, `--redline-test=1500` sets the maximum number of clients to 1500.

> - Upon detecting a request error, the new `FeedbackActor` will pause clients and enter a 30 second sleep period before attempting to scale up again

Reviewer: Is there an option to override the sleep period?

Author: Not currently, but would be a good idea 👀

---
layout: default
title: redline-test
nav_order: 85
parent: Command reference
grand_parent: OpenSearch Benchmark Reference
---

# Redline testing

The `--redline-test` command enables OpenSearch Benchmark to automatically determine the maximum request throughput your OpenSearch cluster can handle under increasing load. It dynamically adjusts the number of active clients based on real-time cluster performance, helping with capacity planning and identifying performance regressions.

> When the `--redline-test` flag is used:

Reviewer 2: ```suggestion
When the `--redline-test` flag is used, OpenSearch Benchmark performs the following steps:
```

> 1. **Client initialization**: OpenSearch Benchmark initializes a large number of clients (default: 1000). You can override this with `--redline-test=<int>`.

Reviewer 2: ```suggestion
1. **Client initialization**: OpenSearch Benchmark initializes a large number of clients (default: 1,000). You can override this with `--redline-test=<int>`.
```

2. **Feedback mechanism**: OpenSearch Benchmark ramps up the number of active clients. A FeedbackActor monitors real-time request failures and adjusts the client count accordingly.

> 3. **Shared state coordination**: OpenSearch Benchmark uses Python’s multiprocessing library to manage shared dictionaries and queues for inter-process communication:

Reviewer 2: ```suggestion
3. **Shared state coordination**: OpenSearch Benchmark uses Python's multiprocessing library to manage shared dictionaries and queues for inter-process communication:
```

   - **Workers** create and share client state maps with the WorkerCoordinatorActor.

>    - **WorkerCoordinatorActor** aggregates client state and forwards it to the FeedbackActor.

Reviewer 2: ```suggestion
   - The **WorkerCoordinatorActor** aggregates client state and forwards it to the FeedbackActor.
```

>    - **FeedbackActor** increases the number of clients until it detects request errors, then pauses clients, waits 30 seconds, and resumes testing.

Reviewer 2: ```suggestion
   - The **FeedbackActor** increases the number of clients until it detects request errors, then pauses clients, waits 30 seconds, and resumes testing.
```

> The following images gives a visual overview of the redline testing architecture:

Reviewer 2: ```suggestion
The following images provides a visual overview of the redline testing architecture.
```

<img src="{{site.url}}{{site.baseurl}}/images/benchmark/osb-actor-system.png" alt="Redline Overview" width="600">

## Usage

To perform a redline test, use the `execute-test` command with the `--redline-test` flag and a timed test procedure.

This test procedure defines a timed workload using the keyword-terms operation. It runs in two phases:

> - **Warmup phase**: The test begins with a warmup period (warmup-time-period) to stabilize performance metrics before measurement begins. This helps avoid skewing results with cold-start effects.

Reviewer 2: ```suggestion
- **Warmup phase**: The test begins with a warmup period (`warmup-time-period`) to stabilize performance metrics before measurement begins. This helps avoid skewing results with cold-start effects.
```

> - **Measurement phase**: During the time-period, OpenSearch Benchmark sends requests at a target-throughput (requests per second) using a specified number of clients. The redline test logic will scale the number of active clients from this baseline to determine the cluster’s maximum sustainable load.

Reviewer 2: ```suggestion
- **Measurement phase**: During the `time-period`, OpenSearch Benchmark sends requests at a `target-throughput` (requests per second) using a specified number of clients. The redline test logic will scale the number of active clients from this baseline to determine the cluster's maximum sustainable load.
```
